textattack/search_methods/metropolis_hastings.py

```diff
@@ -0,0 +1,175 @@
+
+        return scores, model_results
+
+    def _normalize(self, values):
+        """
+        Take list of values and normalize it into a probability distribution
+        """
+        s = sum(values)
+        if s == 0:
+            return [1/len(values) for v in values]
+        else:
+            return [v/s for v in values]
+
+    def _perform_search(self, initial_result):
+        text_len = len(initial_result.tokenized_text.words)
+        max_iter = max(self.max_iter, text_len)
+
+        current_result = initial_result
+        current_text = initial_result.tokenized_text
+        current_score = self._stationary_dist(current_text, initial_result.output)
+        
+        for n in range(max_iter):
+            # i-th word we want to transform
+            i = n % text_len
+
+            transformations = self.get_transformations(
+                        current_text, 
+                        indices_to_modify=[i],
+                        original_text=initial_result.tokenized_text
+                    )
+
+            if len(transformations) == 0:                
+                continue
+
+            scores, model_results = self._batch_stationary_dist(transformations, initial_result.output) 
+            proposal_dist = self._normalize(scores)
+            # Choose one transformation randomly according to proposal distribution
+            jump = np.random.choice(list(range(len(transformations))), p=proposal_dist)
+
+            # Now we have calculate probability of return proposal g(x'|x)
+            reverse_transformations = self.get_transformations(
+                        transformations[jump], indices_to_modify=[i],
+                        original_text=initial_result.tokenized_text
+                    )
+
+            reverse_jump = -1
+            for k in range(len(reverse_transformations)):
+                if reverse_transformations[k].text == current_text.text:
+                    # Transition x -> x' exists
+                    reverse_jump = k
+                    break
+            if reverse_jump == -1:
+                return_prob = 0
+            else:
+                ret_scores, _ = self._batch_stationary_dist(reverse_transformations, initial_result.output)
+                return_prob = self._normalize(ret_scores)[reverse_jump]
+
+            """
+            According to Metropolis-Hastings algorithm
+            let f(x) be value proportional to target distribution p(x)
+            and g(x|x') be transition probability from x' to x.
+            Then, acceptance ratio = min(1, (f(x')*g(x|x')) / (f(x) * g(x'|x)))
+            """
+            if current_score == 0.0:
+                current_score += 1e-9
+            acceptance_ratio = (scores[jump] * return_prob) / (current_score * proposal_dist[jump])
+            acceptance_ratio = min(1, acceptance_ratio)
+            u = np.random.uniform(low=0.0, high=1.0)
+
+            if acceptance_ratio <= u or model_results[jump].succeeded:
```

Comment:
should this say `acceptance_ratio >= u`?

btw, could say `acceptance_ratio >= random.random()` if you `import random`, up to you.

Reply:
critical mistake :(